### DISCUSSION TITLE
Exceptions and JoinableTaskFactory.Run

### DISCUSSION THREAD

**Discussion by Calidus** [microsoft/vs-threading#1339] [Q&A] [answered]
The [docs ](https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.threading.joinabletaskfactory.run?view=visualstudiosdk-2022)include a remark saying 

> Any exception thrown by the delegate is rethrown in its original type to the caller of this method.

When the documentation says rethrown, does that imply that initial exception was caught? I would appreciate some additional context around why this behaves this way and what the intended patterns for dealing with exceptions in the async delegate methods are.

**Comment by AArnott** [comment, marked as answer]
This means that you can catch the exception via its original type, as thrown by the delegate itself. 
The fact that JTF.Run internally catches and rethrows the exception is an implementation detail and something you can ignore. You can think of it as simply allowing the exception thrown from the delegate to propagate to the JTF.Run caller.

**Reply by Calidus** [reply]
Just trying to make sure that I understand, one would expect the following code to catch an exception originally thrown from SomeMethodThatThrowsAsync. 
```
try
{
    joinableTaskFactory.Run(() => SomeMethodThatThrowsAsync());
}
catch (Exception ex)
{
    // do stuff with exception
}
```

Sorry if I am being pedantic, I am attempting to rule out JTF as the cause some exception handling issues I am seeing in a large .NET framework application.

**Reply by AArnott** [reply]
Yes, you can do that. And more precisely, you can do this:
```cs
try
{
    joinableTaskFactory.Run(() => SomeMethodThatThrowsAsync());
}
catch (MySpecificException ex)
{
    // do stuff with exception
}

Task SomeMethodThatThrowsAsync() => throw MySpecificException();
```
